contracts/staking/StakingMain.sol

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: Apache 2.0
+// Copyright © 2022 Artem Belozerov. All rights reserved.
+pragma solidity ^0.8.0;
+
+import "./StakingMain.sol";
+import "./interfaces/IStakingMain.sol";
+import "./interfaces/IStakingEvents.sol";
+import "./common/StakingConstants.sol";
+import "./common/StakingErrors.sol";
+import "./common/StakingStorage.sol";
+import "../common/ownership/Ownable.sol";
+import "../common/interfaces/IERC20.sol";
+import "../voucher/interfaces/IVoucher.sol";
+
+contract StakingMain is IStakingMain, IStakingEvents, Ownable, StakingStorage {
+    function stakeVotes(uint256 votesNumber) external override {
+        uint256 remainder = votesNumber - (votesNumber / StakingConstants.STAKE_STEP) * StakingConstants.STAKE_STEP;
+        if (votesNumber < StakingConstants.STAKE_STEP || remainder != 0) revert StakingErrors.WrongMultiplicity();
+
+        // the first stake must be more then the threshold
+        Stake[] storage stakes = _stakes[msg.sender];
+        if (stakes.length == 0 && votesNumber < _votesThreshold) revert StakingErrors.NotEnoughVotesForStake();
+
+        // check delay between stakes
+        if (stakes.length != 0) {
+            Stake memory lastStake = stakes[stakes.length - 1];
+            if (lastStake.startDate > block.timestamp) revert StakingErrors.StakeDateBeforeNow();
+
+            uint256 timeDelta = block.timestamp - lastStake.startDate;
+            if (timeDelta < StakingConstants.STAKE_COOLDOWN) revert StakingErrors.TooOftenStaking();
+        }
+
+        // check staker balance
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(msg.sender);
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        // append a new stake
+        _pushStake(msg.sender, block.timestamp, votesNumber);
+
+        // TODO: change IERC20 to IVote
+        IERC20(_config.voteAddress).transferFrom(msg.sender, address(this), votesNumber);
+        emit VotesStaked(block.timestamp, votesNumber);
+    }
+
+    function unstakeVotes() external override {
```

Comment:
May run out of gas. Better set some constraints like "start index" & "length" for stakes